src/sql_generator.cpp

```diff
@@ -0,0 +1,312 @@
+#include <iostream>
+#include <string>
+#include <logging.hpp>
+
+#include "../includes/sql_generator.hpp"
+
+using duckdb::DuckDB;
+using duckdb::Connection;
+using duckdb::PreparedStatement;
+using duckdb::DBConfig;
+using duckdb::QueryResult;
+using duckdb::MaterializedQueryResult;
+using duckdb::KeywordHelper;
+
+
+// Utility
+std::string tablename(std::string db_name, std::string schema_name, std::string table_name) {
+    return KeywordHelper::WriteQuoted(db_name, '"')
+           + "." + KeywordHelper::WriteQuoted(schema_name, '"')
+           + "." + KeywordHelper::WriteQuoted(table_name, '"');
+}
+
+void write_joined(std::ostringstream &sql, std::vector<column_def> &columns,
+                  std::function<void(column_def &, std::ostringstream &)> print_col) {
+    bool first = true;
+    for (auto &col: columns) {
+        if (!col.primary_key) {
+            if (first) {
+                first = false;
+            } else {
+                sql << ", ";
+            }
+            print_col(col, sql);
+        }
+    }
+}
+
+void write_joined(std::ostringstream &sql, std::vector<std::string> strings,
+                  std::function<void(std::string &, std::ostringstream &)> print_str) {
+    bool first = true;
+    for (auto &str: strings) {
+        if (first) {
+            first = false;
+        } else {
+            sql << ", ";
+        }
+        print_str(str, sql);
+    }
+}
+
+
+// DuckDB querying
+bool schema_exists(Connection &con, const std::string db_name, const std::string schema_name) {
```

Comment:
Same in all these functions: use references so that string aren't copied.
```suggestion
bool schema_exists(Connection &con, const std::string& db_name, const std::string& schema_name) {
```